% !TEX encoding = UTF-8
% !TEX TS-program = pdflatex
% !TEX root = ../tesi.tex
% !TEX spellcheck = it-IT

%**************************************************************
\chapter{Source Code}
\label{cap:source code}
%**************************************************************

\intro{In this chapter is reported the source code of the algorithms developed by me and used inside the projects}\\

%**************************************************************
\section{Phyton source code}

In these sections are reported all the pyhton sources used in the project
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%5
\subsection{Insertion length}
\tiny
\begin{minted}{python}

class InsertionLengthAlgorithm(BaseAlgorithm):

    def __init__(self, name, path):
        BaseAlgorithm.__init__(self, name, path)

    def run(self):
        self.lengthOfCoverage()
        genome_insertion = []
        sum = 0;
        sumI = 0;

        with open(self.path) as file:
            for line in file:

                if line.find("@") != -1:
                    continue

                pieces = line.split("\t")

                if int(pieces[8]) > 0 and int(pieces[8]) < 10000:
                    isize = abs(int(pieces[3])-int(pieces[7]))
                    genome_insertion.append(isize)
                    sum += isize
                    sumI += 1


        file = "position,isize\n"

        isizeM = int((sum / sumI))
        tmp = 0
        index = 1
        for i in genome_insertion:
            file += str(index)+","+str(i)+"\n"
            dist = (i-isizeM)
            # needs for the standard deviation
            tmp += math.pow(dist,2)
            index += 1


        print "Saving genome_insert_csv.csv file..."
        targetF = open("genome_insert_csv.csv","w")
        targetF.truncate()
        targetF.write(file)
        targetF.close()
        print "Media: "+str(isizeM)
        print "Standard Dev: "+str(format(math.sqrt(tmp/(index)),'.2f'))
        print("Done!")

\end{minted}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%5
\newpage
\subsection{Physical Coverage}
\tiny
\begin{minted}{python}
class PhysicalCoverageAlgorithm(BaseAlgorithm):

    leght = 0;

    def __init__(self, name, path):
        BaseAlgorithm.__init__(self, name, path)

    def run(self):

        self.lengthOfCoverage()
        genome_change = [0] * self.length

        for i in range:
            genome_change.append(0)

        with open(self.path) as file:
            for line in file:
                if line.find("@") != -1:
                    continue;

                pieces = line.split("\t")

                if ((eval(pieces[1]) & 3) == 3) and (int(pieces[8]) > 0):
                    genome_change[int(pieces[3])] = genome_change[int(pieces[3])] + 1

                    genome_change[int(pieces[7]) + len(str(pieces[9]))] = (genome_change[int(pieces[7]) + len(str(pieces[9]))]) - 1

        file = "";
        file += "fixedStep chrom=genome start=1 step=1 span=1\n"
        currentC = 0



        for i in range:
            currentC += genome_change[i];
            file += str(currentC)+"\n"

        print "Saving to physical_coverage.wig "
        targetF = open("physical_coverage.wig","w")
        targetF.truncate()
        targetF.write(file)
        targetF.close()
        print("Done!")
\end{minted}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage
\subsection{Sequence Coverage}
\tiny
\begin{minted}{python}

class SequenceCoverageAlgorithm(BaseAlgorithm):


    def __init__(self, name, path):
        BaseAlgorithm.__init__(self, name, path)

    def run(self):

        self.lengthOfCoverage()
        genome_change = [0] * self.length

        #open the SAM file
        print "Analising SAM file..."
        with open(self.path) as file:
            for line in file:

                #exclude header lines
                if line.find("@") != -1:
                    continue;

                pieces = line.split("\t")

                if ((eval(pieces[1]) & 3) == 3) and (int(pieces[8]) > 0):
                   index = 0

                   for x in xrange(0,len(str(pieces[9]))):
                       genome_change[int(pieces[3])+index] = genome_change[int(pieces[3])+index] + 1
                       index +=1
                else:
                    index = 0
                    if int(pieces[8]) < 0:
                      for x in xrange(0,len(str(pieces[9]))):
                        genome_change[int(pieces[3])-index] = genome_change[int(pieces[3])-index] + 1
                        index +=1



        file = "";
        file += "fixedStep chrom=genome start=1 step=1 span=1\n"

        coverage = 0

        for i in genome_change:
            coverage = genome_change[i];
            file += str(coverage)+"\n"

        print "Saving sequence_coverage.wig file..."
        targetF = open("sequence_coverage.wig","w")
        targetF.truncate()
        targetF.write(file)
        targetF.close()
        print("Done!")
\end{minted}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage
\subsection{Cigar H and S}
\tiny
\begin{minted}{python}
class CigarAlgorithm(BaseAlgorithm):



    def __init__(self, name, path):
        BaseAlgorithm.__init__(self, name, path)

    def run(self):

        self.lengthOfCoverage()
        genome_change = {}

        #open the SAM file
        print "Analising SAM file..."
        with open(self.path) as file:
            for line in file:

                #exclude header lines
                if line.find("@") != -1:
                    continue;

                pieces = line.split("\t")

                #exclude all reads that no have H or S inside CIGAR field
                if pieces[5].find("H") == -1 and pieces[5].find("S") == -1:

                    continue
                else:
                    index = int(pieces[3])
                    length = (int(pieces[3]) + int(pieces[8]))
                    while(index < length ):

                        if genome_change.has_key(index):
                            genome_change[index] = genome_change[index] + 1
                        else:
                            genome_change.__setitem__(index,1)

                        index += 1

        file += "fixedStep chrom=genome start=1 step=1 span=1\n"

        coverage = 0

        for i in genome_change:
            coverage = genome_change[i];
            file += str(coverage)+"\n"

        print "Saving cigar.wig file..."
        targetF = open("cigar.wig","w")
        targetF.truncate()
        targetF.write(file)
        targetF.close()
        print("Done!")
\end{minted}        
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage
\subsection{Kmers Counting}
\tiny
\begin{minted}{python}

class KmersAlgorithm(BaseAlgorithm):

    kmersL = 4
    kmerLabel = ""
    kmerMax = 0

    def __init__(self, name, path):
        BaseAlgorithm.__init__(self, name, path)

    def mapping(self, string):

        string = string.replace("A", "W")
        string = string.replace("T", "A")
        string = string.replace("W", "T")

        string = string.replace("C", "W")
        string = string.replace("G", "C")
        string = string.replace("W", "G")
        return string

    def kmersLength(self):
        print "Please insert the desidered length of the kmers ( leave empty to set default value 4 ) :"
        inputVal = raw_input(">>  ")
        try:
            if inputVal == "":
                self.kmersL = 4
            else:
                print inputVal
                self.kmersL = int(inputVal)
        except ValueError:
            print("Error, you must pass a number")

    def checker(self, max,label):
        if(max > self.kmerMax):
            self.kmerMax = max
            self.kmerLabel = label

    def run(self):

        self.lengthOfCoverage()

        self.kmersLength()
        #open the SAM file

        print "Open SAM file...\n"
        kmers = {}

        print "Looking for kmers and it's opposites for "+str(self.kmersL)+" bases long\n"
        with open(self.path) as file:
            for line in file:

                #exclude header lines
                if line.find("@") != -1:
                    continue;

                pieces = line.split("\t")

                read = str(pieces[9])
                padding = 0
                while len(read[padding:padding+self.kmersL]) >= self.kmersL:
                    key = read[padding:padding+self.kmersL]
                    padding += 1
                    # looking for the kmers
                    if kmers.has_key(key):
                        kmers[key] = kmers[key] +1
                        self.checker(kmers[key],key)
                    # looking for the opposite kmers
                    elif kmers.has_key(self.mapping(key)):
                            kmers[self.mapping(key)] = kmers[self.mapping(key)] +1
                            self.checker(kmers[self.mapping(key)],self.mapping(key))
                    # adding new kmers
                    else:
                        kmers.__setitem__(key,1)
                        self.checker(kmers[key],key)

        file = "kmer,quantity\n"
        for i in kmers:
            file += str(i)+","+str(kmers[i])+"\n"

        print "Saving kmers.csv file..."
        targetF = open("kmers.csv","w")
        targetF.truncate()
        targetF.write(file)
        targetF.close()
        print "The max present value is: "+str(self.kmerMax)+" for the kmer: "+self.kmerLabel+"\n"
        print("Done!")
\end{minted}



\newpage
\section{R source code}
In this section is reported all the R source codes use in the project


\subsection{Insertion length distribution}
\tiny
\begin{minted}{matlab}
chart <- read.csv('../genome_insert_csv.csv', header=TRUE, sep=",")
png('../genome_insert.png')
barplot(
	as.integer(chart$isize), 
	xlim=c(0,20000),
	ylim=c(0,28000) , 
	width = 5,
	main="Insertion length distribution", 
	border = "dark blue", 
	xlab="Isertion length", 
	ylab="Isertions", 
	names.arg = chart$position, 
	horiz = TRUE, 
	col="grey"
)
\end{minted}


\subsection{Kmers distribution}
\tiny
\begin{minted}{matlab}
chart <- read.csv('../kmers.csv', header=TRUE, sep=",")
png('../kmers.png',width = 2000,  height = 600)
barplot(
	as.integer(chart$quantity), 
	xlim=c(0,900), ylim=c(0,15000) , 
	width = 5, 
	main="Kmers distribution", 
	border = "dark blue", 
	xlab="Kmers", 
	ylab="Quantity", 
	names.arg = chart$kmer, 
	horiz = FALSE, 
	col="grey", 
	las=2
)
\end{minted}