% !TEX encoding = UTF-8
% !TEX TS-program = pdflatex
% !TEX root = ../tesi.tex
% !TEX spellcheck = it-IT

%**************************************************************
\chapter{Resequencing}
\label{cap:resequencing}
%**************************************************************

For the project development a bacterial genome was provided\\

This genome belongs to \emph{Lactobacillus casei} and it is quite long genome, 3.079.196 long bases.

The goal of the project is sequence a similar bacterium, defined as sample (or test), using the mate pairs that derived from it, to find and highlight any structural variation as inversions, deletions or insertions.\\
In particular, the tracks are to be produced displayable on Integrative Genome Browser (IGV).
The project includes a part of data analysis to understand the nature and a programming part for the production of results.

%**************************************************************
\section{Instruments}

I used these instruments:
\begin{itemize}
\item IGV that allows the displayment of traks on reference genome.
\item BWA for the alignment of the mate pairs on reference genome.
\item R for the traks plotting
\end{itemize}

For the algorithms development, Python 3 was used.

%**************************************************************
\section{Datas}

For development of the project, the given datas are:\\
    

\begin{itemize}
\item \begin{description}
		\item[Lactobacillus\_caei\_genome.fasta:] the genome of the bacterium
	  \end{description}
\end{itemize}

\begin{itemize}
\item \begin{description}
		\item[lact\_sp\_read1.fastaq:] the first reads of resequencing
	  \end{description}
\end{itemize}

\begin{itemize}
\item \begin{description}
		\item[lact\_sp\_read12.fastaq:] the second reads of resequencing
	  \end{description}
\end{itemize}
  
These last two files contains the reads, the quality of each base of the reads and others useful information that allows the SAM file creation.

From the SAM file I extracted and used some different datas as:

\begin{itemize}
\item \begin{description}
		\item[FLAG:] It provides information related to matching the rid or mate pair. In case of simple alignment without pairing, this value is useful to see if the read is aligned on the positive or negative strand.
	  \end{description}
\end{itemize}


\begin{itemize}
\item \begin{description}
		\item[POS:] position where is aligned the read or mate
		  \end{description}
\end{itemize}

\begin{itemize}
\item \begin{description}
		\item[CIGAR:] how and the kind of the alignement. This field also tell how each base is algined/deleted/inserted or others
  \end{description}
\end{itemize}
\begin{itemize}
\item \begin{description}
		\item[TLEN:] length of mate-pair
		  \end{description}
\end{itemize}
\begin{itemize}
\item \begin{description}
		\item[SEQ:] sequence of bases that make up the read
  \end{description}
\end{itemize}


%**************************************************************
\section{Results}

The results of the project, are composed by wig files ( plain text ), tdf files ( pre digested binary file for IGV, that come from wig files ), csv files ( for statistical manipulation ) and also png file for the data's traks.\\\\

The algorithms, as output, generate the wig files that are converted to tdf files by IGV tools.\\\\

Some results, as the \emph{barchart trak} of the genome insertion lenght, are obtained by passing a csv file, that is generated by algorithm, to a R script that plot and create images.